src/Enumerators/Enumerators.dfy

```diff
@@ -0,0 +1,520 @@
+// RUN: %dafny /compile:0 "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+include "../Frames.dfy"
+include "../Wrappers.dfy"
+include "../Collections/Sequences/Seq.dfy"
+
+module Enumerators {
+
+  import opened Frames
+  import opened Wrappers
+  import opened Seq
+
+
+  // A trait for any value that produces a finite set of value.
+  trait {:termination false} Enumerator<T> extends Validatable {
+    
+    // The Valid() predicate from the Validatable trait ends up
+    // becoming the "enumeration invariant", which in turn becomes
+    // the loop invariant in a while loop that uses an enumerator.
+
+    // All values produced by the Next() method in the order they
+    // were produced.
+    ghost var enumerated: seq<T>
+
+    // Any enumerator that produces one value at a time
+    // and provably terminates is equivalent to an enumerator
+    // that produces a specific seq<T>. This value may be underspecified
+    // such that it is not known, even its length, until after all
+    // values have been produced.
+    // Dafny doesn't let you pass around an underspecified value though,
+    // so we don't define a "to be enumerated" field or function.
+    
+    // The termination measure for the enumerator. Must decrease on every
+    // call to Next().
+    //
+    // Would be better as an arbitrary termination clause somehow instead,
+    // but we don't have language-level access to the built-in well-founded
+    // ordering. See https://github.com/dafny-lang/dafny/issues/762.
+    function Decreases(): nat
+      reads Repr
+      decreases Repr, 1
+      requires Valid()
+
+    // Pre-condition for Next(). Making this a pure predicate means that
+    // enumerators have to at least know ahead of time if they are able to
+    // produce a value, even if they do not know what value it is until
+    // Next() is invoked. This avoids forcing the type parameter for any
+    // enumerator to satisfy the Auto-initializable type characteristic (0),
+    // and allows for much cleaner assertions about state and invariants.
+    // When this is overly restrictive, having an enumerator produce a wrapper
+    // such as Option<T> instead is a good option (ha!), with the caveat that
+    // client code will need to skip over Nones. 
+    predicate method HasNext() 
+      reads Repr
+      requires Valid()
+      decreases Repr, 2
+      ensures HasNext() ==> Decreases() > 0
```

Comment:
You're right, this is redundant since it follows from the decreasing post-condition of `Next()` itself. We need that "extraneous specification" linter :)